connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java

```diff
@@ -0,0 +1,161 @@
+
+    private enum State {
+        INIT, STOPPED, STARTED, FAILED
+    }
+
+    private final String connName;
+    private final ConnectorStatus.Listener statusListener;
+    private final ConnectorContext ctx;
+    private final Connector connector;
+
+    private Map<String, String> config;
+    private State state;
+
+    public WorkerConnector(String connName,
+                           Connector connector,
+                           ConnectorContext ctx,
+                           ConnectorStatus.Listener statusListener) {
+        this.connName = connName;
+        this.ctx = ctx;
+        this.connector = connector;
+        this.statusListener = statusListener;
+        this.state = State.INIT;
+    }
+
+    public synchronized void initialize(Map<String, String> config) {
+        log.debug("Initializing connector {} with config {}", connName, config);
+
+        this.config = config;
+
+        connector.initialize(new ConnectorContext() {
+            @Override
+            public void requestTaskReconfiguration() {
+                ctx.requestTaskReconfiguration();
+            }
+
+            @Override
+            public void raiseError(Exception e) {
+                WorkerConnector.this.state = State.FAILED;
+                ctx.raiseError(e);
+            }
+        });
+    }
+
+    private void start() {
+        try {
+            switch (state) {
+                case STARTED:
+                    return;
+
+                case INIT:
+                case FAILED:
+                case STOPPED:
+                    connector.start(config);
+                    statusListener.onStartup(connName);
+                    this.state = State.STARTED;
+                    break;
+
+                default:
+                    throw new IllegalArgumentException("Cannot start connector in state " + state);
+            }
+        } catch (Throwable t) {
+            log.error("Error while starting connector {}", connName, t);
+            statusListener.onFailure(connName, t);
+            this.state = State.FAILED;
+        }
+    }
+
+    private void pause() {
+        try {
+            switch (state) {
+                case STOPPED:
+                    return;
+
+                case STARTED:
+                    connector.stop();
+                    // fall through
+
+                case INIT:
+                case FAILED:
+                    statusListener.onPause(connName);
+                    this.state = State.STOPPED;
+                    break;
+
+                default:
+                    throw new IllegalArgumentException("Cannot pause connector in state " + state);
+            }
+        } catch (Throwable t) {
+            log.error("Error while shutting down connector {}", connName, t);
+            statusListener.onFailure(connName, t);
+            this.state = State.FAILED;
+        }
+    }
+
+    public synchronized void shutdown() {
```

Comment:
As discussed offline, we will the `stop` method in `Worker` to herder thread and remove the synchronization here.